be/src/storage/update_manager.cpp

```diff
@@ -225,6 +225,68 @@ void UpdateManager::clear_cached_del_vec(const std::vector<TabletSegmentId>& tsi
     }
 }
 
+StatusOr<size_t> UpdateManager::clear_delta_column_group_before_version(KVStore* meta, int64_t tablet_id,
+                                                                        int64_t min_readable_version) {
+    std::vector<std::pair<TabletSegmentId, int64_t>> clear_dcgs;
+    const int64_t begin_ms = UnixMillis();
+    auto is_timeout = [begin_ms]() {
+        if (UnixMillis() > begin_ms + 100) { // only hold cache_clock for 100ms max.
+            return true;
+        } else {
+            return false;
+        }
+    };
+    {
+        std::lock_guard<std::mutex> lg(_delta_column_group_cache_lock);
+        auto itr = _delta_column_group_cache.lower_bound(TabletSegmentId(tablet_id, 0));
+        while (itr != _delta_column_group_cache.end() && !is_timeout() && itr->first.tablet_id == tablet_id) {
+            auto dcg_itr = itr->second.begin();
+            // The delta column group that need to be clear, should satisfy two point:
+            // 1. It's version is under min_readable_version
+            // 2. It's all column are covered by other delta column groups which are also under min_readable_version
+            std::unordered_set<uint32_t> column_set;
+            while (dcg_itr != itr->second.end()) {
+                if ((*dcg_itr)->version() >= min_readable_version) {
+                    // skip the delta column group that isn't expired.
+                    dcg_itr++;
```

Comment:
this may keep an extra version not GCed when there is a version == min_readable_version, also it's better to extract the algorithm and add more UT for that

Reply:
Done